xbmc/platform/win10/filesystem/UwpSMBDirectory.cpp

```diff
@@ -0,0 +1,211 @@
+#include "platform/win32/WIN32Util.h"
+#include "utils/StringUtils.h"
+#include "utils/URIUtils.h"
+#include "utils/log.h"
+#include <string>
+#include <winrt/Windows.Storage.FileProperties.h>
+
+using namespace XFILE;
+using namespace KODI::PLATFORM::WINDOWS;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Foundation::Collections;
+using namespace winrt::Windows::Storage;
+using namespace winrt::Windows::Storage::FileProperties;
+
+bool CUwpSMBDirectory::GetDirectory(const CURL& url, CFileItemList& items)
+{
+  items.Clear();
+
+  auto folder = GetFolder(url);
+  if (!folder)
+    return false;
+
+  std::string path = url.Get();
+  URIUtils::AddSlashAtEnd(path); //be sure the dir ends with a slash
+
+  auto vectorView = Wait(folder.GetItemsAsync());
+  for (auto& item : vectorView)
+  {
+    std::string itemName = FromW(item.Name().c_str());
+
+    CFileItemPtr pItem(new CFileItem(itemName));
+    pItem->m_bIsFolder =
+        (item.Attributes() & FileAttributes::Directory) == FileAttributes::Directory;
+    IStorageItemProperties storageItemProperties = item.as<IStorageItemProperties>();
+    if (item != nullptr)
+    {
+      pItem->m_strTitle = FromW(storageItemProperties.DisplayName().c_str());
+    }
+
+    if (pItem->m_bIsFolder)
+      pItem->SetPath(path + itemName + "/");
+    else
+      pItem->SetPath(path + itemName);
+
+    if (itemName.front() == '.')
+      pItem->SetProperty("file:hidden", true);
+
+    auto props = Wait(item.GetBasicPropertiesAsync());
+
+    pItem->m_dateTime = winrt::clock::to_FILETIME(props.DateModified());
+    if (!pItem->m_bIsFolder)
+      pItem->m_dwSize = static_cast<int64_t>(props.Size());
+
+    items.Add(pItem);
+  }
+
+  return true;
+}
+
+bool CUwpSMBDirectory::Create(const CURL& url)
+{
+  auto folder = GetFolder(url);
+  if (folder) // already exists
+    return true;
+
+  CURL parentUrl = CURL(URIUtils::GetParentPath(url.Get()));
+  folder = GetFolder(parentUrl);
+  if (!folder)
+  {
+    return false;
+  }
+
+  try
+  {
+    std::wstring wStrPath = ToW(url.GetFileNameWithoutPath());
+    Wait(folder.CreateFolderAsync(wStrPath));
+  }
+  catch (const winrt::hresult_error&)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+bool CUwpSMBDirectory::Exists(const CURL& url)
+{
+  return GetFolder(url) != nullptr;
+}
+
+bool CUwpSMBDirectory::Remove(const CURL& url)
+{
+  auto folder = GetFolder(url);
+  if (folder)
+  {
+    try
+    {
+      Wait(folder.DeleteAsync(StorageDeleteOption::PermanentDelete));
+      return true;
+    }
+    catch (const winrt::hresult_error& ex)
+    {
+      std::string error = FromW(ex.message().c_str());
+      CLog::LogF(LOGERROR, __FUNCTION__, "unable remove folder '%s' with error", url.Get(),
+                 error.c_str());
+    }
+  }
+  return false;
+}
+
+StorageFolder CUwpSMBDirectory::GetFolder(const CURL& url)
+{
+  StorageFolder folder{nullptr};
+
+  auto requestedPath = CWIN32Util::SmbToUnc(url.GetRedacted());
```

Comment:
`CURL.GetRedacted()` method returns url which contains USER:PASSWORD instead of actual user name and password, so you shouldn't use it in any place except log messages.

Reply:
Oh, thanks. Now that I look closer at that function I see what you mean. I misinterpreted what GetRedacted meant. My goal was to get a clean SMB path without username and password, i.e. smb://hostname/share/folder -> \\hostname\share\folder.